packages/gatsby/src/utils/page-ssr-module/lambda.ts

```diff
@@ -47,25 +47,132 @@ function setupFsWrapper(): string {
+      applyRename(fsToRewrite, linkedFS, `renameSync`)
+
+      return linkedFS
+    }
+
     // Alias the cache dir paths to the temp dir
-    const lfs = link(fs, rewrites) as typeof import("fs")
+    const lfs = createLinkedFS(fs)
 
     // linkfs doesn't pass across the `native` prop, which graceful-fs needs
     for (const key in lfs) {
       if (Object.hasOwnProperty.call(fs[key], `native`)) {
         lfs[key].native = fs[key].native
       }
     }
-
-    const dbPath = path.join(TEMP_CACHE_DIR, `data`, `datastore`)
-
     // 'promises' is not initially linked within the 'linkfs'
     // package, and is needed by underlying Gatsby code (the
     // @graphql-tools/code-file-loader)
-    lfs.promises = link(fs.promises, rewrites)
+    lfs.promises = createLinkedFS(fs.promises)
+
+    const originalWritesStream = fs.WriteStream
+    function LinkedWriteStream(
+      this: fs.WriteStream,
+      ...args: Parameters<(typeof fs)["createWriteStream"]>
+    ): fs.WriteStream {
+      args[0] = mapPathUsingRewrites(args[0])
+      args[1] =
+        typeof args[1] === `string`
+          ? {
+              flags: args[1],
+              // @ts-ignore there is `fs` property in options doh (https://nodejs.org/api/fs.html#fscreatewritestreampath-options)
+              fs: lfs,
+            }
+          : {
+              ...(args[1] || {}),
+              // @ts-ignore there is `fs` property in options doh (https://nodejs.org/api/fs.html#fscreatewritestreampath-options)
+              fs: lfs,
+            }
+
+      // @ts-ignore TS doesn't like extending prototype "classes"
+      return originalWritesStream.apply(this, args)
+    }
+    LinkedWriteStream.prototype = Object.create(originalWritesStream.prototype)
+    // @ts-ignore TS doesn't like extending prototype "classes"
+    lfs.WriteStream = LinkedWriteStream
+
+    const originalReadStream = fs.ReadStream
+    function LinkedReadStream(
+      this: fs.ReadStream,
+      ...args: Parameters<(typeof fs)["createReadStream"]>
+    ): fs.ReadStream {
+      args[0] = mapPathUsingRewrites(args[0])
+      args[1] =
+        typeof args[1] === `string`
+          ? {
+              flags: args[1],
+              // @ts-ignore there is `fs` property in options doh (https://nodejs.org/api/fs.html#fscreatewritestreampath-options)
+              fs: lfs,
+            }
+          : {
+              ...(args[1] || {}),
+              // @ts-ignore there is `fs` property in options doh (https://nodejs.org/api/fs.html#fscreatewritestreampath-options)
+              fs: lfs,
+            }
+
+      // @ts-ignore TS doesn't like extending prototype "classes"
+      return originalReadStream.apply(this, args)
+    }
+    LinkedReadStream.prototype = Object.create(originalReadStream.prototype)
+    // @ts-ignore TS doesn't like extending prototype "classes"
+    lfs.ReadStream = LinkedReadStream
+
+    const dbPath = path.join(TEMP_CACHE_DIR, `data`, `datastore`)
 
```

Comment:
Seems like there's some repetition here with extending fs.WriteStream and fs.ReadStream, you could consider extracting/reusing.

Reply:
Yes, it's basically the same thing twice (just once for `WriteStream` and once for `ReadStream` - I think it's confusing as it is already and trying to create reusable utility would probably make debugging / understanding it worse than it already is.

Reply:
Doing `class ... extends` would be much more readable, but you can't actually do that is "class" you extend is prototype one like `fs.WriteStream` and `fs.ReadStream` are ( https://github.com/nodejs/node/blob/6b6bcee747c2007117262efb2ff6d61ea888f499/lib/internal/fs/streams.js#L321-L398 being one of them - the other is also in that module)

Reply:
Got it, I'm ok leaving as is.